obp/ope/estimators.py

```diff
@@ -1731,3 +1806,239 @@ def _estimate_mse_score(
+    When the evaluation policy is stochastic, it is not well known whether the balanced IPW performs well.
+
+    Parameters
+    ------------
+    lambda_: float, default=np.inf
+        A maximum possible value of the importance weight.
+        When a positive finite value is given, importance weights larger than `lambda_` will be clipped.
+
+    estimator_name: str, default='b-ipw'.
+        Name of the estimator.
+
+    References
+    ------------
+    Arjun Sondhi, David Arbour, and Drew Dimmery
+    "Balanced Off-Policy Evaluation in General Action Spaces.", 2020.
+
+    """
+
+    lambda_: float = np.inf
+    estimator_name: str = "b-ipw"
+
+    def __post_init__(self) -> None:
+        """Initialize Class."""
+        check_scalar(
+            self.lambda_,
+            name="lambda_",
+            target_type=(int, float),
+            min_val=0.0,
+        )
+        if self.lambda_ != self.lambda_:
+            raise ValueError("lambda_ must not be nan")
+
+    def _estimate_round_rewards(
+        self,
+        reward: np.ndarray,
+        action: np.ndarray,
+        importance_sampling_ratio: np.ndarray,
+        action_dist: np.ndarray,
+        position: Optional[np.ndarray] = None,
+        **kwargs,
+    ) -> np.ndarray:
+        """Estimate round-wise (or sample-wise) rewards.
+
+        Parameters
+        ----------
+        reward: array-like or Tensor, shape (n_rounds,)
+            Reward observed in each round of the logged bandit feedback, i.e., :math:`r_t`.
+
+        action: array-like or Tensor, shape (n_rounds,)
+            Action sampled by behavior policy in each round of the logged bandit feedback, i.e., :math:`a_t`.
+
+        importance_sampling_ratio: array-like or Tensor, shape (n_rounds,)
+            Ratio of probability that the action is sampled by evaluation policy divided by probability that the action is sampled by behavior policy,
+            i.e., :math:`\\hat{\\rho}(x_t, a_t)`.
+
+        action_dist: array-like or Tensor, shape (n_rounds, n_actions, len_list)
+            Action choice probabilities of evaluation policy (can be deterministic), i.e., :math:`\\pi_e(a_t|x_t)`.
+
+        position: array-like or Tensor, shape (n_rounds,), default=None
+            Position of recommendation interface where action was presented in each round of the given logged bandit data.
+            When None is given, the effect of position on the reward will be ignored.
+            (If only one action is chosen and there is no posion, then you can just ignore this argument.)
+
+        Returns
+        ----------
+        estimated_rewards: array-like or Tensor, shape (n_rounds,)
+            Rewards of each round estimated by Balanced IPW.
+
+        """
+        if position is None:
+            position = np.zeros(action_dist.shape[0], dtype=int)
+        iw = (
+            action_dist[np.arange(action.shape[0]), action, position]
+            * importance_sampling_ratio
+        )
```

Comment:
```suggestion
        iw = estimated_importance_weights
```

Reply:
[important] in my understanding after rechecking the reference paper, \hat{\rho} in the paper estimated \pi_e / \pi_b, so we don't have to take the marginalization over the eval policy here.